File: 파이썬 알고리즘 인터뷰/6장/그룹 애너그램.js
// 나의 풀이 (Runtime: 132ms, Memory: 50.5MB)
const groupAnagrams = function(strings) {
  const frequency = {};

  // 각 단어를 배열로 쪼갠 뒤 정렬하여, 애너그램인 경우 동일한 단어로 만든다.
  strings.forEach(string => {
    const key = string.split("").sort().join("");

    // 정렬된 단어를 Key로 사용해, 애너그램을 모은다.
    if (!frequency[key]) {
      frequency[key] = [];
    }

    frequency[key].push(string);
  });

  // 배열로 생성해 반환한다.
  return Object.values(frequency).reduce((acc, curr) => {
    acc.push(curr);
    return acc;
  }, []);
};

File: 파이썬 알고리즘 인터뷰/6장/그룹 애너그램.py
from typing import List


class Solution:
    # 나의 풀이 (Runtime: 96ms, Memory: 17.2MB)
    def groupAnagrams(self, strings: List[str]) -> List[List[str]]:
        frequency = {}
        answer = []

        for string in strings:
            # 문자열을 정렬해 애너그램인 경우, 동일한 문자열로 만든다.
            key = "".join(sorted(string))

            # 해당 문자열을 Key로 사용해 딕셔너리를 구성한다.
            if key in frequency:
                frequency[key].append(string)
            else:
                frequency[key] = [string, ]

        # 구성한 딕셔너리를 순회하며, 정답을 생성한다.
        for key in frequency:
            answer.append(frequency[key])

        return answer